flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/AffinityPropagation.java

```diff
@@ -0,0 +1,535 @@
+	*/
+
+	@SuppressWarnings("serial")
+	private static final class InitAPVertex implements EdgesFunction<Long, Double, Vertex<Long, APVertexValue>> {
+
+		@Override
+		public void iterateEdges(Iterable<Tuple2<Long, Edge<Long, Double>>> edges,
+								Collector<Vertex<Long, APVertexValue>> out) throws Exception {
+
+			Vertex<Long, APVertexValue> APvertexI = new Vertex<>();
+			Vertex<Long, APVertexValue> APvertexE = new Vertex<>();
+
+			Iterator<Tuple2<Long, Edge<Long, Double>>> itr = edges.iterator();
+			Tuple2<Long, Edge<Long, Double>> edge = itr.next();
+
+			APvertexE.setId(edge.f0 * 10 + 1);
+			APvertexE.setValue(new APVertexValue());
+
+			APvertexI.setId(edge.f0 * 10);
+			APvertexI.setValue(new APVertexValue());
+			APvertexI.getValue().getWeights().put(edge.f1.getSource() * 10 + 1, edge.f1.getValue());
+
+			APvertexE.getValue().getOldValues().put(edge.f1.getSource() * 10, 0.0);
+			APvertexI.getValue().getOldValues().put(edge.f1.getSource() * 10 + 1, 0.0);
+
+
+			while(itr.hasNext()){
+				edge = itr.next();
+				APvertexI.getValue().getWeights().put(edge.f1.getSource() * 10 + 1, edge.f1.getValue());
+
+				APvertexE.getValue().getOldValues().put(edge.f1.getSource() * 10, 0.0);
+				APvertexI.getValue().getOldValues().put(edge.f1.getSource() * 10 + 1, 0.0);
+
+			}
+
+			out.collect(APvertexE);
+			out.collect(APvertexI);
+		}
+	}
+
+	/**
+	 * Vertex updater
+	 */
+
+	@SuppressWarnings("serial")
+	public static final class APVertexUpdater extends VertexUpdateFunction<Long, APVertexValue, APMessage> {
+
+		private Long numOfVertex;
+		LongSumAggregator aggregator = new LongSumAggregator();
+
+		public APVertexUpdater(Long numOfVertex){
+			this.numOfVertex = numOfVertex;
+		}
+
+		@Override
+		public void preSuperstep() throws Exception {
+
+			aggregator = getIterationAggregator("convergedAggregator");
+
+		}
+
+		/**
+		 * Main vertex update function. It calls updateIVertex, updateEVertex, computeExemplars or computeClusters
+		 * depending on the phase of the algorithm execution
+		 */
+
+		@Override
+		public void updateVertex(Vertex<Long, APVertexValue> vertex,
+								MessageIterator<APMessage> inMessages) {
+
+			//If all vertices converged compute the Exemplars
+
+			if(getSuperstepNumber() > 1
+				&& (((LongValue)getPreviousIterationAggregate("convergedAggregator")).getValue()
+				== numOfVertex|| getSuperstepNumber() == maxIterations-2)) {
+				computeExemplars(vertex, inMessages);
+				return;
+			}
+
+			//Once the exemplars have been calculated calculate the clusters. The aggregator has a negative value assigned
+			//when exemplars are calculated
+			if(getSuperstepNumber() > 1
+				&& ((LongValue)getPreviousIterationAggregate("convergedAggregator")).getValue()
+				< 0) {
+				if(vertex.getValue().getExemplar() < 0){
+					computeClusters(vertex, inMessages);
+				}
+				return;
+			}
+
+			//Call updateIvertex or updateEvertex depending on the id
+			if(vertex.getId()%2 == 0){
```

Comment:
Are the I and E vertex updates not talking past each other? We had this consideration with HITS where the computation is running twice in parallel and the updates are crisscrossing.